CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
+and
+`Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.
+
+This `AddProof` function will store the proof in a new database `plumoproofdata`, keyed by a unique proof ID.
+
+To propogate these proofs throughout the network, we also propose to extend the `eth` protocol with two new message types: 
+```
+PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`
+GetPlumoProofMsg - Request plumo proof(s) from a peer
```

Comment:
Under what condition would a `GetPlumoProofMsg` be sent to all its peers?
Would be good to have a little more detail here on how exactly the gossiping will work

Reply:
The goal of this msg `GetPlumoProofMsg` is to allow new syncing nodes to fetch previous proofs, and to allow nodes that have gone down to fetch proofs they may have missed.

I'm not sure exactly where this should be handled, but I'm guessing within `downloader`, when a node is syncing, it would send this `get` message.

In regard to sending to all its peers, I imagined sending one message to each peer you know of, and then just taking the union set of all proofs from those responses as your set (in case any peer is similarly out of sync). Maybe this is overkill, if for example peers only respond with `PlumoProofMsg`s if they are in sync already.

Does this address what you were getting at? I'm not very familiar with the already implemented gossiping protocols so appreciate feedback and advice